### DISCUSSION TITLE
Is it possible to get DAE or ODE from a model?

### DISCUSSION THREAD

**Discussion by awecefil** [OpenModelica/OpenModelica#12814] [Q&A] [answered]
Hi, I'm curious if I can get the simplified DAE or ODE of a model? For example, a simple RC circuit with a constant voltage I create by OMEdit like below(I remove the annotations for concise):
```
model test
  Modelica.Electrical.Analog.Sources.SineVoltage sineVoltage(V = 1, f = 60);
  Modelica.Electrical.Analog.Basic.Ground ground;
  Modelica.Electrical.Analog.Basic.Resistor resistor(R = 5);
  Modelica.Electrical.Analog.Basic.Capacitor capacitor(C = 1);
equation
  connect(sineVoltage.n, ground.p;
  connect(sineVoltage.p, resistor.p);
  connect(capacitor.n, resistor.n);
  connect(capacitor.p, sineVoltage.n);
end test;
```
If I run simulation with transformation debugger, I can get all the equations which is ordered. But I am not sure is it possible to get the simplified expression for the DAE or ODE? Take above model as example, the equations are:

![image](https://github.com/user-attachments/assets/bf12099c-4256-4ddf-93b5-5016a6009d6f)

What I want is the equation of der(C.v) after substituting its right hand side by previous equations. Currently, I haven't found a simple/automatic way to get it, I need to do it manually, which is a exhausting work for complicated model.

### **Questions：**

**1. Is it possible to automatically get simplyfied/flatten expression of DAE or ODE from a Modelica model?
2. If 1. is possible, what command in OpenModelica should I use? or other tools can make it?
3. Is model with conditional equation in it also possible to get its DAE or ODE? such as the if-else expressions in some discrete characteristics components**

I have been bothered by this for a while, thus I want to know whether there are some tools match my demands? It's fine to know this isn't possible, I just wanted to confirm it. BTW, I have tried some tools like [pymoca](https://github.com/pymoca/pymoca) which can get the symbolic ODE representation for "very simple Modelica model", there still some works for supporting more MSL components I think.

**Comment by casella** [comment]
@awecefil, you can get the flat DAEs using the [Base Modelica output feature](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/basemodelica.html#flattening-models-to-basemodelica).

In general, computing the derivatives of the DAE by substitution is not a good idea, because you can get very large expressions which are inefficient to compute and also redundant. The computation as laid out in the debugger is the most efficient possible.

Can you make some concrete example with a simple model and the output you'd like to see?

**Reply by awecefil** [reply]
> In general, computing the derivatives of the DAE by substitution is not a good idea, because you can get very large expressions which are inefficient to compute and also redundant. The computation as laid out in the debugger is the most efficient possible.

OK, I can realize that it is inefficient to do substitution. So, when the OpenModelica solver solves the DAE, it may like step-by-step execute the ordered equations?

> Can you make some concrete example with a simple model and the output you'd like to see?

Sure, here I take what [pymoca](https://github.com/pymoca/pymoca) does for example. First, the pymoca takes Modelica code as input and then it will parse the code. After that, a backend generator will generate a python code with using Sympy for the symbolic representation. 
```
# do not edit, generated by pymoca

from __future__ import print_function, division
import sympy
import sympy.physics.mechanics as mech
from pymoca.backends.sympy.runtime import OdeModel
from sympy import sin, cos, tan


class simple_RC_circuit(OdeModel):

    def __init__(self):

        super(simple_RC_circuit, self).__init__()

        # states
        C__v = mech.dynamicsymbols('C.v')
        self.x = sympy.Matrix([C__v])
        self.x0 = {
            C__v : 0.0,
            }

        # variables
        ground__p__v, R__p__v, R__n__v, C__p__v, C__n__v, constantVoltage__p__v, constantVoltage__n__v, R__v, constantVoltage__v, R__i, C__i, constantVoltage__i = mech.dynamicsymbols('ground.p.v, R.p.v, R.n.v, C.p.v, C.n.v, constantVoltage.p.v, constantVoltage.n.v, R.v, constantVoltage.v, R.i, C.i, constantVoltage.i')
        self.v = sympy.Matrix([ground__p__v, R__p__v, R__n__v, C__p__v, C__n__v, constantVoltage__p__v, constantVoltage__n__v, R__v, constantVoltage__v, R__i, C__i, constantVoltage__i])

        # constants
        self.c = sympy.Matrix([])
        self.c0 = {
            }

        # parameters
        R__R, C__C, constantVoltage__V = sympy.symbols('R.R, C.C, constantVoltage.V')
        self.p = sympy.Matrix([R__R, C__C, constantVoltage__V])
        self.p0 = {
            R__R : 5,
            C__C : 0.3,
            constantVoltage__V : 1,
            }

        # inputs
        self.u = sympy.Matrix([])
        self.u0 = {
            }

        # outputs
        self.y = sympy.Matrix([])

        # equations
        self.eqs = [
            ground__p__v - (0),
            R__v - (R__p__v - R__n__v),
            0 - (R__p__i + R__n__i),
            R__i - (R__p__i),
            R__v - (R__i * R__R),
            C__v - (C__p__v - C__n__v),
            0 - (C__p__i + C__n__i),
            C__i - (C__p__i),
            C__i - (C__C * (C__v).diff(self.t)),
            constantVoltage__v - (constantVoltage__p__v - constantVoltage__n__v),
            0 - (constantVoltage__p__i + constantVoltage__n__i),
            constantVoltage__i - (constantVoltage__p__i),
            constantVoltage__v - (constantVoltage__V),
            R__p__v - (C__n__v),
            C__p__v - (ground__p__v),
            constantVoltage__n__v - (R__n__v),
            constantVoltage__p__v - (ground__p__v),
            R__p__i + C__n__i - (0),
            C__p__i + ground__p__i + constantVoltage__p__i - (0),
            constantVoltage__n__i + R__n__i - (0),
            ]

        self.compute_fg()
```
Finally, execute the generated python code, we can get the simplified equations:
![image](https://github.com/user-attachments/assets/0796f345-6831-4e0f-9dd3-a03f5c798ac0)
and the der(C.v) is: 
![image](https://github.com/user-attachments/assets/e03e0d25-6cdb-42e0-8ca9-248575e07cab)
this is totally what I want, the ODE equation of the state. However, pymoca seems not support for more complicated model, thus I want to know if we can get this from OpenModelica?

**Reply by casella** [reply]
> OK, I can realize that it is inefficient to do substitution. So, when the OpenModelica solver solves the DAE, it may like step-by-step execute the ordered equations?

In standard mode, the generated C code solves the DAEs for the derivatives of the states after index reduction, effectively bringing the original DAE model into explicit, state-space form. This is then linked to an ODE solver. Or embedded into an FMU.

If you activate [--daeMode](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-daemode) instead, the system is kept as a DAE and solved with a DAE solver such as IDA. This still undergoes significant processing: high-index systems must be brought to index 1 by symbolic differentiation, and also the DAE solver is only passed the differential equations and the algebraic loops - the equations that can be solved explicitly are solved directly in the generated code and are not explicitly visible to the DAE solver.

**Reply by casella** [reply, marked as answer]
> this is totally what I want, the ODE equation of the state. However, pymoca seems not support for more complicated model, thus I want to know if we can get this from OpenModelica?

No. The problem is, in general you can't get explicit closed-form formulations of the ODEs, for the same reason you can't always get an explicit closed-form solution of a nonlinear equation.

Of course there are cases where this is possible, maybe with the aid of powerful symbolic algebra tools. But that's not what OMC was born to do. The best way to do that is to export the DAEs into a tool like pymoca and continue from there.

You may also check Wolfram System Modeller, which has a Modelica frontend and can use the very powerful Mathematica tool. Unfortunately, this is not open-source software.

**Reply by awecefil** [reply]
Thank you very much for your help!
